### PR TITLE
Added metering area for AF and AE.

### PR DESCRIPTION
**Pull request by pilzus**

### CODE REVIEW

File: library/src/main/api14/com/google/android/cameraview/Camera1.java [hunk 1]
```diff
@@ -398,6 +437,25 @@ private boolean setAutoFocusInternal(boolean autoFocus) {
             return false;
         }
     }
+    private boolean setMeteringRectInternal(Camera.Area area){
+        if(area == null)
```
Comment (kevinpelgrims) [inline library/src/main/api14/com/google/android/cameraview/Camera1.java]: I'd prefer if we use curly brackets for all if statements. Just for the sake of consistency and clarity.

File: library/src/main/api21/com/google/android/cameraview/Camera2.java [hunk 2]
```diff
@@ -414,6 +448,22 @@ private void collectCameraInfo() {
...
+                (CameraCharacteristics.SENSOR_INFO_ACTIVE_ARRAY_SIZE);
+        mSensor.right -= mSensor.left;
+        mSensor.left = 0;
+        mSensor.bottom -= mSensor.top;
+        mSensor.top = 0;
+    }
+    private Rect normalizeMeteringArea (Rect rect){
+
+        // noramalize rect into [0,1]
+        Rect norm = new Rect((rect.left + 1000) / 2,
```
Comment (kevinpelgrims) [inline library/src/main/api21/com/google/android/cameraview/Camera2.java]: Where does +1000 /2 come from?

File: library/src/main/api21/com/google/android/cameraview/Camera2.java [hunk 3]
```diff
@@ -573,8 +661,9 @@ void captureStillPicture() {
             CaptureRequest.Builder captureRequestBuilder = mCamera.createCaptureRequest(
                     CameraDevice.TEMPLATE_STILL_CAPTURE);
             captureRequestBuilder.addTarget(mImageReader.getSurface());
-            captureRequestBuilder.set(CaptureRequest.CONTROL_AF_MODE,
-                    mPreviewRequestBuilder.get(CaptureRequest.CONTROL_AF_MODE));
+            if(getAutoFocus())
```
Comment (kevinpelgrims) [inline library/src/main/api21/com/google/android/cameraview/Camera2.java]: Curly brackets

File: build.gradle [hunk 4]
```diff
@@ -25,7 +25,7 @@ task clean(type: Delete) {
 ext {
     buildToolsVersion = '25.0.1'
     compileSdkVersion = 25
-    minSdkVersion = 9
+    minSdkVersion = 14
```
Comment (pilzus) [inline build.gradle]: Fixed.

File: library/src/main/api14/com/google/android/cameraview/Camera1.java [hunk 5]
```diff
@@ -189,7 +194,24 @@ boolean getAutoFocus() {
             return mAutoFocus;
         }
         String focusMode = mCameraParameters.getFocusMode();
-        return focusMode != null && focusMode.contains("continuous");
+        return focusMode != null && focusMode.contains("continuous") || focusMode.contains("auto");
+    }
+
+    @Override
+    public void setMeteringRect(Rect afRegion) {
+        Camera.Area area = new Camera.Area(afRegion, 1000);
```
Comment (pilzus) [inline library/src/main/api14/com/google/android/cameraview/Camera1.java]: Fixed.

File: library/src/main/api14/com/google/android/cameraview/Camera1.java [hunk 6]
```diff
@@ -214,16 +236,30 @@ void takePicture() {
...
+            // determine current focus mode
+            Camera.Parameters params = mCamera.getParameters();
+            List<String> lModes = params.getSupportedFocusModes();
+            if (lModes != null) {
+                if (lModes.contains(Camera.Parameters.FOCUS_MODE_AUTO)) {
+                    params.setFocusMode(Camera.Parameters.FOCUS_MODE_AUTO); // auto-focus mode if supported
+                    mCamera.setParameters(params);        // set parameters on device
                 }
-            });
-        } else {
```
Comment (pilzus) [inline library/src/main/api14/com/google/android/cameraview/Camera1.java]: Fixed.

File: library/src/main/api14/com/google/android/cameraview/Camera1.java (hunk 1 above)
Comment (pilzus) [inline library/src/main/api14/com/google/android/cameraview/Camera1.java]: Fixed.

File: library/src/main/api21/com/google/android/cameraview/Camera2.java (hunk 2 above)
Comment (pilzus) [inline library/src/main/api21/com/google/android/cameraview/Camera2.java]: This normalises rect dimensions from [-1000, 1000] (Camera1 API) into [0,1000]
The code comment was wrong, I forgot to change it after re-structuring the code.
I have now completely removed the function and updated the code comments.

File: library/src/main/api21/com/google/android/cameraview/Camera2.java (hunk 3 above)
Comment (pilzus) [inline library/src/main/api21/com/google/android/cameraview/Camera2.java]: Fixed.

**Comment by gmnamra** [comment]
the focus rect is specified -600,-500,600,500 with  a comment that says in percent with respect to camera sensor ? What does it actually mean ?

**Comment by gmnamra** [comment]
Is my reading correct that the focus rect is x00 by y00 pixels ?

File: library/src/main/api21/com/google/android/cameraview/Camera2.java (hunk 3 above)
Comment (gmnamra) [inline library/src/main/api21/com/google/android/cameraview/Camera2.java]: Still not clear in pixels what rect is used for measuring focus ? Are there two steps here ? One to select rect within which another rect will be chose for measuring focus hence why the rect size area appears too big ?

**Comment by kevinpelgrims** [comment]
@gmnamra if you have no more comments, I will merge this into master

**Comment by gmnamra** [comment]
Give me this weekend if u can

**Pull request merged**